tesseract/messaging/fees/src/lib.rs

```diff
@@ -451,45 +453,127 @@ impl TransactionPayment {
-			let response_commitment = H256::from_slice(&concat_hash[..32]);
-			let source_key = source.response_commitment_full_key(response_commitment);
-			//Get response receipt keys on dest chain
-			let request_commitment = H256::from_slice(&concat_hash[32..]);
-			let dest_key = dest.response_receipt_full_key(request_commitment);
-			Some((Key::Response { request_commitment, response_commitment }, source_key, dest_key))
-		});
+		let (req_errors, requests): (Vec<_>, Vec<_>) = req_results
+			.into_iter()
+			.filter_map(|x| x)
+			.partition(|res| res.is_err());
+
+		let zero_fee_keys: Vec<_> = req_errors.into_iter().filter_map(|e| e.err()).collect();
+		let requests: Vec<_> = requests.into_iter().filter_map(|e| e.ok()).collect();
+
+		let resp_results: Vec<_> = stream::iter(response_entries)
+			.then(|data| {
+				let source = source.clone();
+				let dest = dest.clone();
+				async move {
+					// Get response commitment keys on source chain
+					let concat_hash = match hex::decode(data.hash.clone()) {
+						Ok(bytes) => bytes,
+						Err(_) => return None,
+					};
+
+					let response_commitment = H256::from_slice(&concat_hash[..32]);
+					//Get response receipt keys on dest chain
+					let request_commitment = H256::from_slice(&concat_hash[32..]);
+
+					match source.query_response_fee_metadata(request_commitment).await {
+						Ok(fee) if fee.is_zero() => {
+							Some(Err(Key::Response { request_commitment, response_commitment }))
+						}
+						Ok(_) => {
+							let source_key = source.response_commitment_full_key(response_commitment);
+							let dest_key = dest.response_receipt_full_key(request_commitment);
+							Some(Ok((
+								Key::Response { request_commitment, response_commitment },
+								source_key,
+								dest_key,
+							)))
+						}
+						Err(err) => {
+							tracing::warn!(
+                            "Failed to query response fee metadata for request {:?}: {:?}",
+                            request_commitment,
+                            err
+                        );
+							let source_key = source.response_commitment_full_key(response_commitment);
+							let dest_key = dest.response_receipt_full_key(request_commitment);
+							Some(Ok((
+								Key::Response { request_commitment, response_commitment },
+								source_key,
+								dest_key,
+							)))
+						}
+					}
+				}
+			})
+			.collect()
+			.await;
+
+		let (resp_errors, responses): (Vec<_>, Vec<_>) = resp_results
+			.into_iter()
+			.filter_map(|x| x)
+			.partition(|res| res.is_err());
+
+		let zero_fee_keys: Vec<_> = zero_fee_keys
+			.into_iter()
+			.chain(resp_errors.into_iter().filter_map(|e| e.err()))
+			.collect();
+		let responses: Vec<_> = responses.into_iter().filter_map(|e| e.ok()).collect();
 
 		let mut keys_to_delete = vec![];
 		let mut keys_to_prove = vec![];
 
-		for key in requests.chain(responses) {
+		for key in requests.into_iter().chain(responses) {
 			if hyperbridge.check_claimed(key.0.clone()).await? {
```

Comment:
Yeah, just query fee metadata here and push keys with zero fee metadata into the keys to be deleted